first draft of parameter design doc

tfoote opened the pull request:
Connects to #57

tfoote committed:
first draft of parameter design doc

adolfo-rt commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+When a parameter changes value it should be possible to get a notification.
+
+### Reject parameter changes
```
If I understand correctly, this would be the method for enforcing parameter ranges (among other things). And 'be able to communicate the preconditions' would refers to a mechanism that could be used to, for instance, display valid parameter values or ranges, as `rqt_reconfigure` currently does. Is my interpretation correct?

tfoote replied:
Exactly.

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+- Gonzalo's research on parameters.
+  - Discussion: https://groups.google.com/forum/#!topic/ros-sig-ng-ros/YzCmoIsN0o8 and https://groups.google.com/forum/#!searchin/ros-sig-ng-ros/parameter/ros-sig-ng-ros/fwDBcei5Ths/L6ORPfjUDXYJ
+  - Protype: https://github.com/abellagonzalo/dynamic_config
```
prototype (not protype)

dirk-thomas committed:
fix spelling

dirk-thomas commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+### Return parameter values
+
+Also at the core functionality return the value of one or more parameters which have previously been set.
```
_Return_ as well as _unset_ need to support groups of parameters atomically if _setting_ those is supported.

tfoote replied:
Added a line after to clarify it will return `unset` explicitly.

dirk-thomas replied:
My comment was referring to the need to query a set of parameters atomically. The new sentence does not yet cover that.

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+## Ideal System
+
+It is useful to consider the ideal system to understand how it relates to the current system and how a new system could work.
+We'd like to support both the use cases of standard parameters as well as dynamic parameters.
```
Can you define standard parameters in this section? Are standard parameters just ROS 1 parameters?

dirk-thomas commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+- `int64`
+- `string`
+- `bool`
+- `bytes[]`
```
This looks like an arbitrary list. Why is there no array of any type beside `bytes`?

tfoote replied:
We chose non-complex non-custom types. And do not provide the full complement of things like unsigned to support scripting.  I'd prefer not to support `bytes` but when we reviewed the API it was deemed important to support binary blobs of which bytes[] was the default choice. 

In our discussions support for arrays was discussed and deferred deciding on adding support for later discussions. I'll add this to the open questions below.

dirk-thomas replied:
I think the support of arrays of e.g. `int64` is crucial to consider in the article. A lot of existing code sets array of primitive data on the parameter server.

tfoote replied:
Added paragraph on arrays at bottom of the article for things to address.

BobDean replied:
what is the reasoning behind not supporting 16 and 32 bit values?  Is it assumed that the host language or client access api will take care of the conversion?

tfoote replied:
In general shorter bit values are important for space and speed optimization. However supporting those optimizations increases the complexity of the interface. In this case it would close to double the number of datatypes supported. People would then have to be careful about up and down conversions and range checks which becomes noteably harder when interacting with loosely typed languages.

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+### Introspection of expected parameters
+
+When using parameters providing a list of expected parameters can prevent accidentally setting parameters with typos in their names etc.
```
Another grammar nitpick: comma after "When using parameters". The sentence is a bit hard to parse in its current state.

actually, don't worry about these comments, I'll address the little things in my own branch.

dirk-thomas committed:
fix spelling

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+### Logging and playback
+
+When logging an entire system the parameter changes can be logged via standard topic recording of the events channel.
```
What is meant by an "entire system"? The entire ROS 2 graph? Or can it be a subset of the nodes involved in getting/setting parameters?

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+## Topics not covered in at the moment
+
+There are still topics to discuss and flesh out either in this document or others going forward a few to highlight.
```
I don't really understand the last few words of this sentence. Should it be "Going forward, there are still topics to discuss and flesh out either in this document or others. A few to highlight:"

Commit:
A few grammar fixes

jackieokay commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+
+There are still topics to discuss and flesh out either in this document or others going forward a few to highlight.
+
+## Parameter initialization
```
Is this supposed to be a subsection of "Topics not covered at the moment"? It should probably have 3 leading `#`s.

esteve commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,184 @@
+Currently there a few parts of the specification unimplemented.
+
+- No parameter subscription registration.
+  The events are published, but there is not way to register a callback for changes to a parameter.
```
There is the `on_parameter_event` method.

https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/parameter_client.hpp#L232

tfoote committed:
address comments from @dirk-thomas @esteve and @jaquelinekay

tfoote committed:
atomic reads, formatting, datatype clarifications

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+We'd like to support both the use cases of ROS 1.0 built in parameters as well as dynamic parameters.
+An ideal parameter system would have the qualities laid out in the following paragraphs.
+
+### Accept parameter values
```
I'd call this section "set" and the next "get" parameter values. This makes it line up with "unset" and it avoids "accept" which implies that it could have not been accepted ("reject"), which is not covered in the following paragraph.

tfoote committed:
use set and get from @wjwwood

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+
+### Reject parameter changes
+
+If a parameter change is requested to an invalid value, the parameter server should be able to reject the change.
```
I'd say "the owner of the parameter should be able to reject ...", using "parameter server" implies how it should be implemented since server invokes the idea of a centralized client-server relationship.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+
+### Logging
+
+All changes to parameters should be visible when logging.
```
"... should be loggable."?

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+### Logging
+
+All changes to parameters should be visible when logging.
+And when playing back a log file the parameter changes should be able to be reapplied.
```
I'm not sure about this line, since it's possible that any change that is played back could be rejected. Maybe it would be more appropriately scoped to say that it should be possible to playback parameter events.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+
+To cover the feature set above, the ROS 2.0 parameter system is proposed as follows.
+
+### Parameters Hosted on nodes
```
"Parameters hosted by nodes" or "Parameters Hosted by Nodes"? Maybe "in" instead of "by" or "on".

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+- `bool`
+- `bytes[]`
+
+The datatypes are chosen as non-complex primatives.
```
`primitives`, and `non-complex` seems poorly defined to me, i.e. what does that mean?

dirk-thomas replied:
The terms are used in the interface article: http://design.ros2.org/articles/interface_definition.html

Maybe we should reference the article somewhere here?

wjwwood replied:
Based on http://design.ros2.org/articles/interface_definition.html I'd say that `non-complex primitive` is redundant? Since that article describes a `non-primitive` message as complex?

dirk-thomas replied:
I would agree to that.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+Each node is responsible for providing the following functionality.
+
+- **Get Parameters**
+  Given a list of parameter names it will return the parameter values.
```
Should this be "given a node and a list of parameter names" or "a list of parameter addresses"?

tfoote replied:
This is the functionality the "node" is providing. I was assuming this had already been addressed to the node, vs the client API which requires the node name and parameter name to first address the right node.

wjwwood replied:
Well, then maybe a sentence just before `- **Get Parameters**` which states something to this affect. I think that would help since we don't differentiate here between a "core api" and a "client api". It's not clear to me whether or not what's described here is for the local API or the remote, ROS based, API. And if the descriptions are agnostic, then it shouldn't assume the node name is implied, since both ways don't have it implicitly.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+- **Set Parameters**
+  Given a list of parameter names, it will request an update of the values subject to validation of the values.
+  The updated values can include unsetting the value.
+  It will provide an API that can atomically update a set of values such that if any of the values fail validation, none of the values are set.
```
I think it would be clearer to have `get`, `get atomic`, `set`, and `set atomic` all called out separately, rather than making notes about atomic versions in the description.

tfoote replied:
This is pushing toward API level choices like below. I don't think it's necessary to require that there's both a get and get atomic. For example the regular get can be atomic.

BobDean replied:
this is something i did not understand when reading the code, why there is an atomic and a non-atomic api when just atomic would suffice.

wjwwood replied:
@BobDean It's more like atomic and batch. With atomic, no values are set if any are rejected. With non-atomic some values may be set while others are rejected. Basically non-atomic could be replaced with many calls to the atomic version, one for each value, but that would be less efficient.

wjwwood replied:
After reading my comment again, `get atomic` may not make sense. From my perspective, `get atomic` just means give me the current state of several values without worrying about part of them getting updated while I'm getting them.

wjwwood replied:
I guess my latest comment isn't that clear either, let me try one more time:

What I meant by `get atomic` is it gets the value of several parameters with a guarantee that they're not being updated at the same time. What I mean by "may not make sense" is that you could argue that all gets should be atomic, therefore removing the need for a specifically atomic get.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+  Given a list of parameter names, it will request an update of the values subject to validation of the values.
+  The updated values can include unsetting the value.
+  It will provide an API that can atomically update a set of values such that if any of the values fail validation, none of the values are set.
+  The success or failure of this call will be available to the client.
```
Should this mention that the non-atomic set will return a success/failure for each value and that the atomic will only return one success/failure? Also should we mention the failure reason is included or is that too much of an implementation detail?

tfoote replied:
I think that's getting too detailed for the current form of this article. Once we have this we can then drive down to the API specifics.

BobDean replied:
i recommend:
-   the response of the set operation (local or remote) will only be related to the node accepting or rejecting the set request. It will not include a concept of how the change may or may not have affected the behavior of the system. +   Verification callbacks should be designed to execute as quickly as possible, and not include information external to the parameter metadata, with the exception of managed lifecycle state.

"external to"  is meant to imply that a node may reject a parameter due to being in the wrong lifecycle state. ideallly the "metadata" would include enough information to validate the value.   Other than these two cases, what are valid reasons to reject the set request?  If there are edge cases, i believe they would be the responsibility of system architecture, and not a parameter api.

tfoote replied:
I think limiting the verification to the specific metadata is overly limiting. It's not the best example, but you could say that a request to change a serial driver to a non-extant device could be rejected. Of course that depends on if your driver can support running and polling for a non-existant device to become available, in which case it might be ok to configure it for a non-extant device. 

There are other things on the system which are not possible to become available that parameters might reflect such as the number of cores, or a core to which you want affinity for. Or the GPU device id. 

Many of these things can be covered by what I talked about in the other comment with querying for a set of valid inputs, instead of just the conditions. And thinking about this more I see it as generalizing that we want to be able to get a `usage:` style description for any parameter. And more generally this will apply to the whole Node API with services and topics with descriptions and potentially dynamic validation ranges.

I'm going to update this section to try to clarify from @BobDean and @wjwwood.

gbiggs replied:
Doing things like checking for hardware presence when validating a parameter is introducing complex functionality into the handling of parameters. I'm not arguing for whether this is right or wrong - it's a system design issue, in my opinion, so the capability should probably be there. However, at least for managed nodes, it needs to be restricted. If the node is active and allows parameter changes while active, then validation probably should be restricted as @BobDean suggests to avoid impacting execution schedules. Changes that require more involved validation would then be restricted to being performed while the node is inactive.

I guess what this typically poorly-worded comment is saying is that there may need to be a distinction between parameters that can be deterministically (in time) set and parameters that cannot, with a restriction on when the latter can be changed.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+- **List Parameters**
+  Provide a list of parameter names which are currently set.
+- **Describe Parameters**
+  Given a list of parameter names, return their datatype.
```
Doesn't this return more than the data type? I believe it should also include meta information like bounds and stuff that would be used programmatically by a tool like the `dynamic_reconfigure_gui`.

tfoote replied:
Right now it only returns the datatype. Early prototypes had a lot more details in this descriptor but the problem is that requires the core api to be updated any time the validation features change or update. I'd rather see the validation preconditions etc be done via a separate channel.

BobDean replied:
recommendation that there is metadata that is part of the description, but is not required for validation. example: comment describing what the parameter does (so parameters are self documenting), the source of the parameter (config file, another node, command line...). 
- Given a list of parameter names, return their descriptive metadata including datatype, description, last update time, and source of last update.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+### Backwards compatibility Parameter Server like behavior
+
+There are use cases where the older behavior with parameters persisting beyond the duration of a specific node are valuable.
+To this end we propose to write a simple node which emulates the policy of the ROS 1.0 parameter server: it runs in namespace `/` and simply accepts all change requests.
```
I think this might be mixing two issues. One is the desire to have a global parameter space where no one owns the parameters, i.e. they are not validated by anyone. The other is to have parameters persist when the owner is shutdown and restarted, which can apply to either parameters owned and validated by a node or to a global parameter server.

wjwwood commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,200 @@
+
+Each node will provide a topic on which parameter changes will be published.
+This topic is to support monitoring parameters for change.
+It is expect that client libraries will implement the ability to register callbacks for specific parameter change notifications using this topic.
```
Should this be "parameter events" rather than "parameter changes"? Events would encompass: accept (atomic?) change, reject (atomic?) change, request (atomic?) change?, get value?, request unset (might be covered by request change), and reject unset. I'm not sure how thorough this document should be.

tfoote replied:
I'm updating it to use parameter events nomenclature. The current implementation only publishes events of the changes. If you want the raw requests and potential rejections hopefully you can log the actual requests themselves by logging the RPC calls.

wjwwood commented:
I had several comments and suggestions, but I don't think any are controversial or blockers, so +1 after they are addressed.

tfoote committed:
more from @wjwwood

BobDean commented:
what does a parameter being 'set' by the service mean (i.e. i get back a SetParameterResult of true)? that the value is now in the local cache? Or that the behavior of the node has changed as a result? 
(splitting into multiple posts for tracking)

wjwwood commented:
> what does a parameter being 'set' by the service mean (i.e. i get back a SetParameterResult of true)? that the value is now in the local cache? Or that the behavior of the node has changed as a result? 
> (splitting into multiple posts for tracking)

I think that must mean that the parameter has been set. Whether or not the parameter has changed the behavior of the node depends on how the user write their code. But it's an important distinction.

I could imagine that you have a few possible types of callbacks the user could provide (not exhaustive and not all maybe be needed or a good idea):
- Validate, which gets called for an individual parameter.
  - Validate batch, which gets called when a group of parameters are changed (maybe grouping is done by namespace, I don't know)
- On update, called anytime a parameter is set and accepted.
- On change, called anytime a parameter is set and accepted and is different from the previous value.
  - Both this and on update could have other settings like coalescing or throttling, and both could have batch versions.

The validate and on update could be collapsed into each other, where on update returned true or false for whether or not the change was accepted.

If the user changes the behavior in the validate callback, then a successful set would indicate the node's behavior changed. If they did so in an on change or separate on update callback then a successful set may not indicate node behavior has changed. Since we cannot control when or where the user makes use of the parameter then I think we cannot tie that to the successful result of a set call.

BobDean commented:
From the cpp code, it seems to me that you are thinking of a parameter as a being set of separate entities, vs an entity with multiple parts (struct/class). We had a similar split api as the proposed.  After using it for a while, the realization was that in most cases we were usually making all of calls to gather all of the metadata.  We then switched to an api which worked on a single ConfigValue type for the network aspects. 

So for rframe the equilvalent of ParameterValue, ParameterType, ParameterDescriptor etc is a single ConfigValue data type:

PrimitiveEnum type // bool, string, unit, int etc
string value // value are stored as strings, arguable w.r.t. efficiency
string comment // a descriptive comment, required in our api. 
string  source // the name of whoever or whatever set the value last, source code, config file, gui, great for debugging
timestamp ts  // change timestamp, ns since epoch

the basic api then is:
get(paramName,param,defaultValue,comment)  // sets the value if it does not exist, retrieves it if id does (i.e. loaded from a config file)
get(paramName,value)  // retrieve a value which _should_ exist. returns NOT_FOUND (unset), CHANGED or NOT_CHANGED
set(paramName,value)  // set existing param, fails if not found, side effect: changes the source and timestamp values
keyMap() // returns the parameter set as a map<paramName,configValue>
count(paramName)  // number of duplicate keys, we allow them as xml and other markup formats allow them.

get/set then have versions for handing vector<type> to get multiple instances, why we support that is another discussion.

our executor then has a callback for a ConfigurationChange, which sets the local value from an external source.  we do not currently have "sets" of parameters, but i am thinking about it after looking at the rclcpp api. if there is a race condition related to a parameter change, we handle using other techniques (currently).

BobDean commented:
i agree that what occurs within the set is an important distinction. in my opinion the set should return
very quickly, which implies that it should not result in a change to node
behavior. What is 'ok' for that change to be is a wide range of
possibilities connected to LifeCycle and safe operation of the robot, which
may take a while to process and not immediately return.

Validation is a solid point. If there was a way to handle it within the
parameter api, that would be my preference (i think someone mentioned this
above).  maybe when the value is initialized a valid parameter range is
passed in. or a lambda.  in other words, using a callback for validation is
fine, but i do not see a need to include parameter validation in a callback
group.

(also, why can i not get my comment to trail the post i am responding to?)

On Thu, Sep 17, 2015 at 4:07 PM, William Woodall notifications@github.com
wrote:

> what does a parameter being 'set' by the service mean (i.e. i get back a
> SetParameterResult of true)? that the value is now in the local cache? Or
> that the behavior of the node has changed as a result?
> (splitting into multiple posts for tracking)
> 
> I think that must mean that the parameter has been set. Whether or not the
> parameter has changed the behavior of the node depends on how the user
> write their code. But it's an important distinction.
> 
> I could imagine that you have a few possible types of callbacks the user
> could provide (not exhaustive and not all maybe be needed or a good idea):
> - Validate, which gets called for an individual parameter.
>   - Validate batch, which gets called when a group of parameters are
>     changed (maybe grouping is done by namespace, I don't know)
> - On update, called anytime a parameter is set and accepted.
> - On change, called anytime a parameter is set and accepted and is
>   different from the previous value.
>   - Both this and on update could have other settings like coalescing
>     or throttling, and both could have batch versions.
> 
> The validate and on update could be collapsed into each other, where on
> update returned true or false for whether or not the change was accepted.
> 
> If the user changes the behavior in the validate callback, then a
> successful set would indicate the node's behavior changed. If they did so
> in an on change or separate on update callback then a successful set may
> not indicate node behavior has changed. Since we cannot control when or
> where the user makes use of the parameter then I think we cannot tie that
> to the successful result of a set call.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros2/design/pull/58#issuecomment-141211548.

tfoote commented:
A parameter being set means that it's value has been updated in the node's parameter storage and that all callbacks registered related to the parameter update validation have returned true. We cannot make guarentees about the behavior of the node but if you query the parameter from anywhere, local or remotely it will be the new value (assuming you are querying before another change has made).

As @wjwwood mentioned it is up to the implementer to use the parameter values. If the code has latency in taking in the parameter, such as it will reprocess parameters on the next cycle for a looping process, the parameter is still accepted and set. I don't think it would be a good idea to recommend that the parameter setting block until the next cycle once the parameter has begun to be acted upon. You can take it a step further and not just ask has the node changed behavior but has the hardware changed behavior. Lets take the case that you're setting the desired velocity for a spinning sensor which has a large inertia. If it's spinning at 10 Hz and you ask it to go to 15Hz. The main thread is running in isolation, reading the target velocity once per revolution. Accelerating from 10Hz to 15Hz will take 1-2 seconds. With this setup there are three different times you could report that the parameter has been successfully set. You could report it once the value is validated and recorded, once the value has been passed to the other thread or once the system has completed it's change to the requested value. 

To be able to track whether a goal point has been achieved I think it's improper to try to use the return code from the parameter set call as a way to measure success. The feedback on the state of the system should come from a separate channel such as the measured velocity from an encoder being streamed on a topic. Similarly it's quite possible that if there's a low battery voltage or other side effect out of the purview of the node which makes a perfectly valid target RPM is unachievable. In this case it would be a bad idea for the set parameter call to blocked until achievement of the goal. 

Since we've determined that waiting a potentially indeterminate amount of time for the hardware to react, I think that the same can be said for the software. In the case that I outlined it would be possible for the callback to block until the active thread notified it that it had read the new value and was operating upon it. However that's still a very indeterminate value as that thread may be blocked or do something odd or improper and never actually send the command to the hardware. Thus giving a false sense of "doing something". Thus I would suggest that if introspection of the current set point is important that it should be instrumented for streaming feedback via topic.

Another reason for not relying on the callbacks to imply that the "node is acting on my request" is that a successful callback does not guarantee the state of the remote node either as anther process might have made a different successful request immediately following your request. You could also register for callback on change of the parameters. 

Similarly if the set parameter calls block pending execution or partial execution you also need to start getting into determining queuing behavior and potentially preemption such that two requests coming in almost simultaneously will cause the first one to "fail" as it's value was overriden before the next cycle, or else it queues up and then gets applied in the following cycle as the first callback was still waiting for the full cycle. As I describe this it sounds more like we're getting into the territory of actions instead of parameters. It seems that we should set the policy to returning quickly once the input has been validated. 

I see you replied while I was composing this, and you agree. Since you asked it's clearly not obvious. How much should be added to the document to clarify?

BobDean commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,202 @@
+
+There are several ways to load parameters at startup including command line arguments, roslaunch arguments, and potentially parameter files.
+This is something which should be addressed in conjunction with the new launch system.
+
```
the more 'required' bits of metadata at initialization the better, as it helps with self documentation. i recommend that initialization occur at a single location in the node.

BobDean commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,202 @@
+
+### Parameter update validation
+
+The node will allow for the registration of a callback for custom parameter value validation.
```
has thought to how to ship the validation around occurred? or if it is needed? The valdiation parameters could be bound to the callback.  Given that parameter updates are rare in the system messaging lifetime, having a "validateParameter" api call which validates the value on the node without setting it could enable a dynamic configure like api to work without needing to ship around the validation metadata, or require that validation meet a well defined interface.

wjwwood replied:
You've sort of asked this in more than one place, but validation happens in the node which owns it and can be arbitrarily complex. The metadata is only provided so that simple bounds can be placed on input so that graphical tools can render widgets for the user to use. For example, the Kinect driver might have a parameter for resolution and frequency, which are type int. The metadata would indicate that only values 0-2 are valid for each of those parameters and that each value corresponds to some helper text (like the corresponding resolution or frequency). However, the validation maybe more sophisticated, because the user might select two independently valid selections like high resolution and high rate, but the node can still reject them because that combination of parameters is not valid. This part of the validation is not captured in a machine readable format in the meta data, so it isn't used in the parameter setting UI, but is checked when trying to set the parameters.

tfoote replied:
Allowing a query for in-depth validation without applying is something we could consider. But I don't think that replaces communicating the validation outward. It's very valuable for the guis to be able to have a concept of valid values to be able to render things like slider bars or drop down lists. There are certainly things which cannot be validated remotely. For example checking on local assets. However we could consider exposing a service where the gui could query for prevalidated ranges/values. So that when a gui is pointed at parameter foo. It will call the parameter server and ask get_foo_valid_inputs and it returns something expressing the range [-1,1] for a float to be rendered as a slider or a list of detected serial ports for a drop down device configuration in the GUI. I think this would also be a better place to consider including the description as the description is also something which is useful to users but not to other nodes. It could be rendered next to the slider or other visual element.

tfoote committed:
clarify core api

BobDean commented:
There are two implications that I see from the design which I do not see directly discussed:

1) There is an implication that parameters may be set while the robot is in an unsafe state. Where unsafe is described as one where changes may cause unexpected results, or cause a failure which may result in a safety problem for nearby humans and objects. Versus putting the robot into a safe state, changing parameters, and resuming the behavior. An example of where safety procedures overlap with aspects of the api and could replace parts of the api. If agreed, then text should be added that parameters may be set in an unsafe state, and it is then the responsibility of the node to deal with that change in a safe manner. 

2) There is an implication that multiple controllers may be changing parameter state at the same time. Use cases describing how this may occur would be useful to further discussions on how to protect against such a case.

wjwwood commented:
> 1) There is an implication that parameters may be set while the robot is in an unsafe state. Where unsafe is described as one where changes may cause unexpected results, or cause a failure which may result in a safety problem for nearby humans and objects. Versus putting the robot into a safe state, changing parameters, and resuming the behavior. An example of where safety procedures overlap with aspects of the api and could replace parts of the api. If agreed, then text should be added that parameters may be set in an unsafe state, and it is then the responsibility of the node to deal with that change in a safe manner.

+1 to more clearly state this in the design doc. From my perspective this is a more advanced behavior for parameters. So just plain parameters shouldn't have any constraint on when they can be set, but the node can always validate it. When a node has life cycle imposed on it (we've been calling this a managed node, but that's up for discussion), then we can also enforce that only a "managed parameter" can be used with the "managed node". This "managed parameter" can hook in before the user's validation callback for all parameter sets and ensure that the parameter is not changed when in the wrong state. Therefore these nodes will require you to bring them out of the active state before changing a parameter. Regardless of the details of when or why parameter changes will be automatically rejected, I think it should be built on top of basic parameters, not into them.

BobDean commented:
My concern was primarially with node behavior safety. The "setting in an
unsafe state" is more tied to how the parameter set interacts with the
running behavior, which I feel is separate from lifecycle.  It is important
to lifecycle, but it should be possible for someone to write a "safe" node
without it.  In other words, the set parameter call is a "request" to the
node. The node should respond to it as the designer feels is appropriate.
(which I believe we all agree on, yes?)

On Fri, Sep 18, 2015 at 4:14 PM, William Woodall notifications@github.com
wrote:

> 1) There is an implication that parameters may be set while the robot is
> in an unsafe state. Where unsafe is described as one where changes may
> cause unexpected results, or cause a failure which may result in a safety
> problem for nearby humans and objects. Versus putting the robot into a safe
> state, changing parameters, and resuming the behavior. An example of where
> safety procedures overlap with aspects of the api and could replace parts
> of the api. If agreed, then text should be added that parameters may be set
> in an unsafe state, and it is then the responsibility of the node to deal
> with that change in a safe manner.
> 
> +1 to more clearly state this in the design doc. From my perspective this
> is a more advanced behavior for parameters. So just plain parameters
> shouldn't have any constraint on when they can be set, but the node can
> always validate it. When a node has life cycle imposed on it (we've been
> calling this a managed node, but that's up for discussion), then we can
> also enforce that only a "managed parameter" can be used with the "managed
> node". This "managed parameter" can hook in before the user's validation
> callback for all parameter sets and ensure that the parameter is not
> changed when in the wrong state. Therefore these nodes will require you to
> bring them out of the active state before changing a parameter. Regardless
> of the details of when or why parameter changes will be automatically
> rejected, I think it should be built on top of basic parameters, not into
> them.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros2/design/pull/58#issuecomment-141555124.

tfoote commented:
Yes I agree that any chance of parameter is a request and the node is allowed to reject it for "safety" reasons or any other policy from the designer. 

Sort of along the same lines as William was suggesting I think it would be good for us to have some recommended policies/patterns which relate to the managed nodes and lifecycles potentially with wrappers which implement that for you.

BobDean commented:
I may have missed it, but it there an ability to share a "global" or
"common" parameter? such as a common url for a resource, or operational
policy which effects multiple nodes.

On Sat, Sep 19, 2015 at 9:46 PM, Tully Foote notifications@github.com
wrote:

> Yes I agree that any chance of parameter is a request and the node is
> allowed to reject it for "safety" reasons or any other policy from the
> designer.
> 
> Sort of along the same lines as William was suggesting I think it would be
> good for us to have some recommended policies/patterns which relate to the
> managed nodes and lifecycles potentially with wrappers which implement that
> for you.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros2/design/pull/58#issuecomment-141724639.

tfoote commented:
@BobDean that's in the paragraph [Backwards compatibility Parameter Server like behavior](https://github.com/ros2/design/pull/58/files#diff-4075c1e78328ec132523517becd96e1bR151)

jack-oquin commented:
Aren't global parameters more than just a "backwards compatibility" concern?

The concept seems useful in its own right. Some parameters like `use_sim_time` and the old `tf_prefix` are inherently global, they do not apply to just one node.

mikeferguson commented:
> The concept seems useful in its own right. Some parameters like use_sim_time and the old tf_prefix are inherently global, they do not apply to just one node.

Or, even more common: the robot_description URDF parameter.

tfoote commented:
I think that the unowned parameters are a pattern that generally should be discouraged. The example of the `robot_description` is a good example where it should have a more clearly defined interface to deal with changing values. https://github.com/ros/robot_state_publisher/issues/29

There are very few pieces of data which are truely global and unchaning like `use_sim_time`.

jack-oquin commented:
How will unowned parameters be discouraged?

Fairly often, I find it useful to define parameters at a higher level in the namespace for use by more than one node. Isn't that what `searchParam()` is for?

In most cases, there can be a node that "owns" each of those parameters. But, what happens if that node should terminate unexpectedly? Will the parameter persist across a respawn? It would in ROS 1.

Instead of dismissing them as a backwards compatibility option, I think global parameters deserve some serious design treatment.

wjwwood commented:
> Fairly often, I find it useful to define parameters at a higher level in the namespace for use by more than one node. Isn't that what searchParam() is for?

@jack-oquin It would be fine to have something like `searchParam()` which works in combination with the "list params" function we described here. One of the things we've avoided discussing so far is addressing in ROS 2. I think there is a need for a more formal addressing scheme in ROS, but that's a topic for another article. The only reason I bring it up is that it should be possible to have hierarchies of nodes and parameters within namespaces and a tool like `searchParam()` will still be useful.

> In most cases, there can be a node that "owns" each of those parameters. But, what happens if that node should terminate unexpectedly? Will the parameter persist across a respawn? It would in ROS 1.

I know that's a useful feature that a lot of people use and we've talked about how to preserve that behavior if it is desired, but it's not mentioned explicitly here either as a use case or as something that needs addressing.

@tfoote we should add that point, or I can if you don't have time.

My personal opinion is that for ROS 2 it should be done by a tool which is actually a normal node. This node would act similar to how it's done in ROS 1, but would not use any special interfaces or features if possible. In ROS 1 a node starts up, then contacts the parameter server and gets the current values (either defaults or what was set from a previous run). The generic parameter server node in ROS 2 could provide the same service, which nodes could optionally query before taking the hard coded defaults in their code. Like the ROS 1 parameter server, this generic parameter server node in ROS 2 could provide a global namespace which all nodes could reference and search through and could hold common global parameters like `robot_description` (though I've long thought that should be a latched topic in ROS 1 so it could be captured by a bag file). In both cases parameters are lost if the parameter server is shutdown or crashes and in both cases nodes could store parameters in it and recover them after the node crashes or restarts.

So I think that even if all parameters must be owned by a node, i.e. no true global parameters, you can model this pattern using nodes. We just need to provide the right tools and we could include an instance of this parameter server node in something like the `roscore` command.

> Instead of dismissing them as a backwards compatibility option, I think global parameters deserve some serious design treatment.

I agree, we should describe the motivation for this feature and describe how it can be accomplished with the proposed solution and let people comment on that.

tfoote committed:
updating parameter server details and adding mention of searchparam

tfoote commented:
I've added some more details about how to use a parameter server node in parent namespaces and added a section on `searchParam` to address the above comments.

wjwwood committed:
reworked the ideal system section to be more concise, have less of an impact on the TOC

tfoote committed:
Merge pull request #66 from ros2/parameter_design_style

reworked the ideal system section to be more concise, have less of an impact on the TOC

tfoote committed:
validator method added to ideal system.

dirk-thomas committed:
wording

dirk-thomas committed:
clarify notifications

dirk-thomas committed:
add notice about incrementally listing parameters

dirk-thomas commented:
The bullet `Check what parameters are expected to be set for a particular Node` listed under `Ideal System` doen't comes from the capabilities in ROS 1. I don't see a reason to consider this necessary for an "ideal system". Having a node prepopulate all its valid parameters and reject others seems to be sufficient. The valid parameter names can also be exposed as "constants" to other entities (if they don't want to hard code the values into their code). Therefore I would propose to remove that specific bullet (and the later remark that it hasn't been implemented anyway).

If the part should stay in the article (which I think it shouldn't) it would be better suited in the "Topics not covered at the moment" section at the end.

dirk-thomas committed:
note about persistence

dirk-thomas commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,207 @@
+
+### Parameter update validation
+
+The node will allow for the registration of a callback for custom parameter value validation.
```
The wording of this sentence sounds very fuzzy to me. It could be interpreted that external entities can register these validation callbacks which I think is not the case?

dirk-thomas commented on articles/055_ros_parameter_design.md:
```diff
@@ -0,0 +1,207 @@
+
+### Parameter API
+
+The client libraries will provide an API for interfacing with the Core Parameter API for both local and remote nodes including return codes.
```
This sentence seems to say almost the same as the last paragraph of "Required functionality" (https://github.com/ros2/design/pull/58/files#diff-4075c1e78328ec132523517becd96e1bR141).

dirk-thomas commented:
Beside my comments I added a few commits to the PR rephrasing little things and adding a some additional information.

tfoote commented:
@BobDean 

> (also, why can i not get my comment to trail the post i am responding to?)

Github doesn't support branching "threading" of replies. If you want to reply to something specific the best thing to do is to quote it and then reply. You get specific sub-threads from line comments, but they're kind of volatile as if the line of code is updated the whole sub-thread can vanish.

tfoote committed:
Addressing feedback from @dirk-thomas

tfoote commented:
I've addressed @dirk-thomas 's comments. I'd like to suggest we merge this now and get it posted on the site for others to find. I'm sure there will be more updates to come but it's at a relatively self consistent state at the moment.

tfoote committed:
clarify node validates parameters

tfoote committed:
style cleanup re #67

tfoote commented:
Thanks everyone for the feedback. It's now on the site at: http://design.ros2.org/articles/ros_parameters.html
